Documentation for building native executables with Mandrel

galderz opened the pull request:
Closes #11032

@zakkak @jerboaa @karm can you have a read through the changes?

galderz committed:
Documentation for building native executables with Mandrel #11032

zakkak commented on docs/src/main/asciidoc/building-native-image.adoc:
```diff
@@ -14,7 +14,50 @@ This guide covers:
+== GraalVM
+
+Building a native executable requires using a distribution of GraalVM.
+There are two distributions: Oracle GraalVM and Mandrel.
```
There are actually 3 distributions. GraalVM CE, GraalVM EE, and Mandrel.

zakkak commented on docs/src/main/asciidoc/building-native-image.adoc:
```diff
@@ -14,7 +14,50 @@ This guide covers:
+* A working container runtime (Docker, podman)
+* The code of the application developed in the link:getting-started[Getting Started Guide].
+
+Skip to the <<#container-runtime,this section>> to continue with the guide for Mandrel.
```
Maybe highlight that the user should consult the mandrel-related tip?

jerboaa commented:
Looks OK to me.

gsmet commented on docs/src/main/asciidoc/building-native-image.adoc:
```diff
@@ -327,6 +370,14 @@ You can also select the container runtime to use with:
+Building with Mandrel requires a custom builder image parameter to be passed additionally:
+[source,shell]
+----
+./mvnw package -Pnative -Dquarkus.native.container-build=true -Dquarkus.native.builder-image=quay.io/quarkus/ubi-quarkus-mandrel:20.1.0.1.Alpha2-java11
```
I don't think we want that hardcoded here. It should be in the build-parent with the GraalVM version.

galderz replied:
Yeah, I wasn't happy about the hardcoding either. Even if it's in the build-parent, it'd be a hardcode of some sort. 

I wonder if we should have a a `20.1` quay tag that just links to the latest 20.1 release. That'd make it more palatable. @Karm @zakkak?

jerboaa replied:
+1

galderz replied:
Created https://github.com/graalvm/mandrel/issues/95

gsmet commented on docs/src/main/asciidoc/building-native-image.adoc:
```diff
@@ -14,7 +14,50 @@ This guide covers:
+
+* Mandrel is only available for projects that can be compiled with Java 11.
+If you are using Java 8,
+you should consider using Oracle GraalVM instead.
```
I read all these points and I only see drawbacks for Mandrel: it has less features, doesn't have some enhancements and doesn't support Java 8. It's a bit of a hard sell :).

Karm replied:
/me chipping in from my vacation: The fact that something is *smaller* and *focused* on a particular function is IMHO a huge plus. I am not a copywriter, yet I think we should formulate something to the fact that Mandrel is minimized and focused. In the age of bloatware where a simple web page pushes 10MB of data you didn't want on you, there is a value in delivering less.

zakkak replied:
+1 we should point out that mandrel distribution is ~200MB while graalvm CE is ~400MB.

adinn replied:
@gsmet This is not at all about us providing an inferior product. It is about us providing a better level of support to the majority of our users.

mandrel removes a bundle of GraalVM features that are not needed for the vast majority of Quarkus native apps. It omits OpenJDK 'enhancements' to JDK11 that our OpenJDK team does not manage and cannot vouch for (that's especially important when it comes to conformance and security). It does not attempt to work on ljdk8 but that's a legacy release of OpenJDK which most Quarkus users will have no good reason to adopt and for which, in consequence, we have no commitment to support Quarkus beyond best effort. In other words, mandrel streamlines and simplifies the build, test, release and support cycle for most of our users, especially those who pay for support.

Note that we are not forcing anyone's hand here. We won't refuse to support anyone using Quarkus on GraalVM for preference whether with a labs jdk11u or labs jdk8u. Indeed, we will do or best to provide support for any problems that manifest in GraalVM when using that configuration. Obviously, we would much prefer if our paid support customers were to use mandrel. If they don't then should any problems manifest on GraalVM we will almost certainly want them to check -- where possible -- that those  problems also reproduce on mandrel. If they cannot then we will still try to help but we are not going to write a blank cheque based on whatever code Oracle labs might add to GraalVM beyond the native image component. Anyway, the idea that we are pushing some inferior variant at the expense of GraalVM is very far from the truth.

gsmet commented on docs/src/main/asciidoc/building-native-image.adoc:
```diff
@@ -327,6 +370,14 @@ You can also select the container runtime to use with:
+----
+./mvnw package -Pnative -Dquarkus.native.container-build=true -Dquarkus.native.builder-image=quay.io/quarkus/ubi-quarkus-mandrel:20.1.0.1.Alpha2-java11
+----
+Check https://quay.io/repository/quarkus/ubi-quarkus-mandrel?tab=tags[here] to find out about available tags.
```
You haven't closed your TIP.

gsmet reviewed (changes requested):
Added some comments.

jaikiran commented on docs/src/main/asciidoc/building-native-image.adoc:
```diff
@@ -14,7 +14,50 @@ This guide covers:
+If you are using Java 8,
+you should consider using Oracle GraalVM instead.
+
+* Mandrel is currently only recommended for containerized environments.
```
Hello @galderz, is this sentence implying that the `native-image` *generation* needs to happen within a container or is this saying that the built native image needs to be run within a container?
Either way, I'm curious about the requirement of a containerized environment. In the context of Mandrel, is there some specific advantages that containerized environments provide? Are these advantages the same as probably Graal VM too provides in containerized environment?

jerboaa replied:
It's not required. But it's convenient. If you don't want containerized, you can just install a mandrel release from here and run with it (by pointing GRAALVM_HOME to the install): https://github.com/graalvm/mandrel/releases

We say that, because on Mac and Windows you can use the container to produce linux native images and deploy on Linux as no Windows and Mac community builds exist.

We only provide Linux builds so far. If somebody steps in to do Mac and Windows community builds even better. It probably won't be us doing those (in the foreseeable future). Note that for Mac builds one would have to do OpenJDK 11u vanilla mac builds too or use some other base JDK which would open up the Mac build to a different class of problems.

galderz replied:
@jaikiran I'm expanding the points here to make it clearer.

jaikiran replied:
The updated section looks much more clearer to me. Thank you @galderz

galderz commented:
Addressed all feedback and pushed an update.

galderz commented:
Hold off integrating this yet, I'm further enhancing the section that explains the differences between Mandrel and Oracle GraalVM CE/EE.

zakkak commented on docs/src/main/asciidoc/building-native-image.adoc:
```diff
@@ -14,7 +14,54 @@ This guide covers:
+* A working container runtime (Docker, podman)
+* The code of the application developed in the link:getting-started[Getting Started Guide].
+
+Skip to the <<#container-runtime,this section>> to continue with the guide for Mandrel,
```
I think this should be 
```suggestion
Skip to <<#container-runtime,this section>> to continue with the guide for Mandrel,
```

galderz replied:
It's fine as is, the text on the right is what is shown.

zakkak replied:
Exactly, so the end result will be `Skip to the this section to...` while I think it should be `Skip to this section to...`

galderz replied:
Ah yes, you're suggesting removing `the`. Thought you meant the opposite :)

galderz commented:
The PR is ready once again. I've added a little bit more information on the reasons behind the Mandrel differences. Also fixed @zakkak latest comment.

jerboaa commented on docs/pom.xml:
```diff
@@ -108,6 +108,8 @@
                         <graalvm-version>${graal-sdk.version-for-documentation}</graalvm-version>
                         <graalvm-flavor>${graal-sdk.version-for-documentation}-java11</graalvm-flavor>
+                        <mandrel-version>${mandrel.version-for-documentation}</mandrel-version>
+                        <mandrel-flavor>${mandrel.version-for-documentation}-java11</mandrel-flavor>
```
Is this intentional? Wouldn't this expression, `${mandrel.version-for-documentation}-java11`, evaluate to `20.1.0.1.Alpha2-java11-java11`?

galderz replied:
Yeah, nicely caught!

galderz replied:
Pushed update with fix.

jerboaa commented on docs/src/main/asciidoc/building-native-image.adoc:
```diff
@@ -14,7 +14,61 @@ This guide covers:
+
+* Mandrel is built slightly differently to Oracle GraalVM CE, using the standard OpenJDK project.
+This means that it does not profit from a few small enhancements that Oracle have added to the version of OpenJDK used to build their own GraalVM downloads.
+This enhancements are omitted because the Red Hat OpenJDK team does not manage them, and cannot vouch for.
```
Suggestion: ".... omitted because upstream OpenJDK does not manage them and, thus, cannot vouch for them." It's not only Red Hat who's working on 11u OpenJDK upstream.

galderz replied:
Pushed update with fix.

jerboaa commented on docs/src/main/asciidoc/building-native-image.adoc:
```diff
@@ -14,7 +14,61 @@ This guide covers:
+
+* less than 15 minutes
+* an IDE
+* JDK 11 installed with `JAVA_HOME` configured appropriately
```
This seems misleading. The Mandrel release comes with OpenJDK 11 bundled. So it doesn't need OpenJDK 11 just Mandrel, unless somebody wants to build it from source, which I don't think this guide is about. Am I missing something?

jerboaa replied:
I wonder if this should be `JAVA_HOME` or `GRAAL_VM_HOME`. I'm thinking the latter pointing to the mandrel install, no?

zakkak replied:
The guide is about using the builder image. So there is nothing to point `GRAALVM_HOME` to.

zakkak replied:
Or to put it more clearly, you won't have mandrel on your system to point `JAVA_HOME` to it. You will only have the `ubi-quarkus-mandrel` image for native compilation part.

jerboaa replied:
Apparently it's not clear what this doc is referring to and what those reqs are for. Should we spell it out? I'm guessing the JDK 11 install is for compiling the quarkus app?

galderz replied:
Not sure there's much we can do here. The sources get compiled by the java version outside the container. If you're on Windows/macOS you can set a JAVA_HOME to Mandrel cos there's no distribution for that. On linux you could download Mandrel and set Java home to that (in the absence of fedora RPMs...etc), but that's as far as you can say. So, with all these caveats I'd probably leave it as is for now.

zakkak commented on build-parent/pom.xml:
```diff
@@ -34,6 +34,7 @@
         <!-- The Graal version we suggest using in documentation - as that's
            what we work with by self downloading it: -->
         <graal-sdk.version-for-documentation>20.1.0</graal-sdk.version-for-documentation>
+        <mandrel.version-for-documentation>20.1.0.1.Alpha2-java11</mandrel.version-for-documentation>
```
Just noticed that @jerboaa already commented on this :) https://github.com/quarkusio/quarkus/pull/11061#discussion_r466923014

galderz commented:
Pushed update with fixes to latest review comments.

jerboaa reviewed (approved):
LGTM

galderz commented:
@gsmet This seems ready now. Can you integrate?

galderz committed:
Documentation for native executable debugging #11032

galderz commented:
This has not yet been integrated, anything wrong?

In the mean time, I've added an extra section with information on debugging native executables. Please have a look.

zakkak commented:
Since this is not integrated yet, could we please wait for the Final image to be made public first? So that the instructions can point to it instead of the alpha?

@cescoffier said the image will go public later today (see https://github.com/quarkusio/quarkus-images/pull/88#issuecomment-674786690)

galderz commented:
@zakkak If we have the floating 20.1 tag, we don't have to wait for Final image :)

zakkak commented:
> @zakkak If we have the floating 20.1 tag, we don't have to wait for Final image :)

Correct, but we still have to wait for the floating 20.1 tag, which we need to ask @cescoffier to create for us.

cescoffier commented:
`20.1.0.1.Final-java11`is not available from Quay - https://quay.io/repository/quarkus/ubi-quarkus-mandrel?tab=tags

zakkak commented:
> `20.1.0.1.Final-java11`is not available from Quay - https://quay.io/repository/quarkus/ubi-quarkus-mandrel?tab=tags

I guess you mean "is now available".

Thanks @cescoffier

galderz commented:
Thx @cescoffier. I'll update the PR for final.

Once we have a 20.1 floating tag we can send a docu update.

galderz commented:
Updated to final and pushed.
